soc/xtensa/intel_adsp/common/soc_mp.c

```diff
@@ -312,28 +314,21 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 
 	z_mp_stack_top = Z_THREAD_STACK_BUFFER(stack) + sz;
 
-#ifdef CONFIG_IPM_CAVS_IDC
-	idc = device_get_binding(DT_LABEL(DT_INST(0, intel_cavs_idc)));
-#endif
-
-	/* Enable IDC interrupt on the other core */
-	idc_reg = idc_read(IPC_IDCCTL, cpu_num);
-	idc_reg |= IPC_IDCCTL_IDCTBIE(0);
-	idc_write(IPC_IDCCTL, cpu_num, idc_reg);
-	/* FIXME: 8 is IRQ_BIT_LVL2_IDC / PLATFORM_IDC_INTERRUPT */
-	sys_set_bit(DT_REG_ADDR(DT_NODELABEL(cavs0)) + 0x04 +
-		    CAVS_ICTL_INT_CPU_OFFSET(cpu_num), 8);
-
 	/* Send power up message to the other core */
 	uint32_t ietc = IDC_MSG_POWER_UP_EXT((long) z_soc_mp_asm_entry);
 
-	idc_write(IPC_IDCIETC(cpu_num), 0, ietc);
-	idc_write(IPC_IDCITC(cpu_num), 0, IDC_MSG_POWER_UP | IPC_IDCITC_BUSY);
+	IDC[curr_cpu].core[cpu_num].ietc = ietc;
+	IDC[curr_cpu].core[cpu_num].itc = IDC_MSG_POWER_UP | IPC_IDCITC_BUSY;
 
-	/* Disable IDC interrupt on other core so IPI won't cause
-	 * them to jump to ISR until the core is fully initialized.
+#ifndef CONFIG_SOC_SERIES_INTEL_CAVS_V25
+	/* Early DSPs have a ROM that actually receives the startup
+	 * IDC as an interrupt, and we don't want that to be confused
+	 * by IPIs sent by the OS elsewhere.  Mask the IDC interrupt
+	 * on other core so IPI won't cause them to jump to ISR until
+	 * the core is fully initialized.
 	 */
-	idc_reg = idc_read(IPC_IDCCTL, cpu_num);
+	uint32_t idc_reg = idc_read(IPC_IDCCTL, cpu_num);
+
 	idc_reg &= ~IPC_IDCCTL_IDCTBIE(0);
 	idc_write(IPC_IDCCTL, cpu_num, idc_reg);
```

Comment:
is this the desired approach now, that sometimes IDC registers are accessed via `idc_read()` and `idc_write()` and sometimes via `IDC[src_core].core[dst_core].x` or is there a well defined rule when to use which approach?

Reply:
The rule is "use the new API", but I've left the older usage in place in a bunch of places out of conservatism.  In this case the delta is just a noop -- I removed the declaration of the same variable above this line so needed to add the "uint32_t " here.